### ISSUE TITLE
getopts should use grapheme clusters for text alignment 

### ISSUE THREAD

**Issue by steveklabnik** [rust-lang-nursery/getopts#7]
<a href="https://github.com/Kimundi"><img src="https://avatars.githubusercontent.com/u/2903206?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [Kimundi](https://github.com/Kimundi)**
_Saturday Mar 23, 2013 at 13:52 GMT_

_For earlier discussion, see https://github.com/rust-lang/rust/issues/5516_

_This issue was labelled with: A-libs, E-easy in the Rust repository_

---

It should at minimum align to number of codepoints (characters), (and now that PR #8710 has landed, it should be doing that minimum), but the only correct thing would be to align to grapheme clusters.

test case

```
extern mod extra;
use extra::getopts::groups;

fn main() {
    let optgroups = ~[
        groups::optflag("a", "apple",        "apple description"),
        groups::optflag("b", "banana\u00AB", "banana description"),
        groups::optflag("c", "br\xfbl\xe9e", "br\xfbl\xe9e quite long description"),
        groups::optflag("k", "kiwi\u20AC",   "kiwi description"),
        groups::optflag("o", "orange\u2039", "orange description"),
        groups::optflag("r",
                        "raspberry-but-making-this-option-way-too-long",
                        "raspberry description is also quite long indeed longer than every other piece of text we might encounter here and thus will be automatically broken up"),
    ];
    println(usage("Usage: fruits", optgroups));
}

```

current output:

```
% rustc /tmp/issue5516.rs && /tmp/issue5516 --help
warning: no debug symbols in executable (-arch x86_64)
Usage: fruits

Options:
    -a --apple          apple description
    -b --banana«        banana description
    -c --brûlée         brûlée quite long descripti
    -k --kiwi€          kiwi description
    -o --orange‹        orange description
    -r --raspberry-but-making-this-option-way-too-long 
                        raspberry description is also quite long indeed longer
                        than every other piece of text we might encounter here
                        and thus will be automatically broken up


```

Note that in addition to things failing to line up, it is also cutting off the end of the text for the brûlée case.  That issue should also be fixed.

**Comment by SimonSapin** [comment]
Perhaps `StrExt::width` is more relevant here than grapheme clusters.

**Comment by tromey** [comment]
The original example here looks correct.  However I was able to see the bug by using a different character in one of the options, in my case \u{3042}.

Also:

> Note that in addition to things failing to line up, it is also cutting off the end of the text for the brûlée case. That issue should also be fixed.

This doesn't seem to happen any more.

**Comment by prataprc** [comment]
This is interesting issue. Never thought about grapheme length in rendering context. Is it possible to solve this issue in a clean way for terminal emulators ?

**Comment by SimonSapin** [comment]
> Perhaps `StrExt::width` is more relevant here than grapheme clusters.

This is now https://crates.io/crates/unicode-width

**Comment by prataprc** [comment]
@SimonSapin  Looks like there are two versions to compute the display width. `width()` and `width_cjk()` how to know which context to use ? Is this related to localisation ?

**Comment by SimonSapin** [comment]
Many CJK characters are typically two columns wide in monospace fonts. The crate above will tell you, based on the Unicode character database. But some characters have an "ambiguous width": they make take 2 or 1 column depending on whether you’re in a "CJK context", which is not all that well defined. In the Windows console, it appears to depend on the Windows locale. The difference between these two methods is how ambiguous characters are counted. https://www.unicode.org/reports/tr11/ has all of the background.

But the actual rendering will depend on a bunch of stuff including which fonts are installed. Still, just using `width()` and ignoring the locale is already a better approximation than `.chars().count()`.

**Comment by prataprc** [comment]
One additional note, In the PR #71 I am using crate attribute: `#![feature(rustc_private)]` Got some strange error
```
use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead? (see issue #27812)
```

What does this error message mean ? Should we use the `rustc_private` feature ?

Thanks,

**Comment by SimonSapin** [comment]
No, you should probably not use unstable features in this crate since that would prevent using it with Rust Stable, whereas previous versions can.

**Comment by KodrAus** [comment]
This was implemented in #71 and released as `0.2.18`.
